### ISSUE TITLE
Cluster Docs

### ISSUE THREAD

**Issue by linvon** [uber/cadence#3499]
Is there any document about cadence server cluster?
I want to use Cadence now, but I didn't see any documentation on high availability and load balancing, how does it implement these features, and how do I configure them?
The most important thing is that we are not using k8s and docker, so i want to know how to configure and deploy cadence-server, any one can help?

**Comment by linvon** [comment]
Another question is if i deploy Cadence NDC Service, will there be n public ips or only one ip? 
How do i configure client if there is more than one public cadence server ip?

**Comment by longquanzheng** [comment]
> Another question is if i deploy Cadence NDC Service, will there be n public ips or only one ip?
> How do i configure client if there is more than one public cadence server ip?

Looks like similar to https://github.com/uber/cadence/issues/3464

**Comment by longquanzheng** [comment]
> Is there any document about cadence server cluster?
> I want to use Cadence now, but I didn't see any documentation on high availability and load balancing, how does it implement these features, and how do I configure them?
> The most important thing is that we are not using k8s and docker, so i want to know how to configure and deploy cadence-server, any one can help?

The only thing you need for high availability is to scale up the cluster size for Frontend, and also history/match accordingly as your workload increase. There is no particular LB required for connecting to Cadence frontend, because FE is stateless service, you could use any LB mechanism. 

The only doc I can see is https://github.com/uber/cadence/blob/master/docker/README.md#quickstart-for-production
But welcome to any contribution. As I believe there are a lot of people not using K8s.

**Comment by linvon** [comment]
> The only thing you need for high availability is to scale up the cluster size for Frontend, and also history/match accordingly as your workload increase. There is no particular LB required for connecting to Cadence frontend, because FE is stateless service, you could use any LB mechanism.

Thanks a lot, 
But i noticed that in NDC deployment, requests to no active FE will be autoforward to active FE if AutoForwarding is enable, and won't be handled if not enable. So i think there is no different between request direct to active FE and use LB.
Therefore i have another question, the only way i find to switch active FE is use CLI command manually, when active FE is down, requests won't be handled until active FE restart, so how does HA work? Can not it switch active FE automatically? I think now the only use of NDC is to make sure that the task will not lost when AutoForwarding is enable

**Comment by longquanzheng** [comment]
>Therefore i have another question, the only way i find to switch active FE is use CLI command manually, when active FE is down. requests won't be handled until active FE restart

 HA is mostly enough for a single cluster, because there are many FE nodes in a a cluster. So when a FE node is down, there should be no impact for availability.

XDC is for even higher availability needs, in some cases like DC failure. In this case you might want to failover to a cluster that lives in a healthy DC. In this case, Cadence CLI can help you do failover. You can also integrate with a DC health monitor service, if you have one, if you want to have automatic failure experience.

**Comment by linvon** [comment]
Thanks for reply！

> HA is mostly enough for a single cluster, because there are many FE nodes in a a cluster. So when a FE node is down, there should be no impact for availability.

One last question I want to ask, does a single cluster integrate multiple FE without additional configuration? Or need to specify multiple FE in the configuration file? Cause I didn't find any configuration about this

**Comment by longquanzheng** [comment]
> Thanks for reply！
> 
> > HA is mostly enough for a single cluster, because there are many FE nodes in a a cluster. So when a FE node is down, there should be no impact for availability.
> 
> One last question I want to ask, does a single cluster integrate multiple FE without additional configuration? Or need to specify multiple FE in the configuration file? Cause I didn't find any configuration about this

Yes you need to specify `ringpop seeds` to let all the nodes know each others so that they are connected as a cluster. 
The trick is in this config in the doc that I sent you.
```
    -e RINGPOP_SEEDS=10.x.x.x,10.x.x.x  \               -- csv of ipaddrs for gossip bootstrap
```
If you use 
```
make bins 
```
command to build Cadence project, you will get a `cadence-server` binary and then you should be able to run the command without Docker. 


See here if you want to understand how ringpop works: https://github.com/uber/ringpop-go . I don't think you have to unless you run into some issues in the future.

**Comment by linvon** [comment]
Thank you very much！
